lldb_bson.py

```diff
@@ -1,236 +1,1449 @@
-        return
+    This class implements that cache of memory segments
+    """
 
-    process = debugger.GetSelectedTarget().GetProcess()
-    frame = process.GetSelectedThread().GetFrameAtIndex(0)
+    def __init__(self):
+        self._segments: dict[int, bytes] = {}
+        "Segments of memory keyed by the base address of the read operation"
 
-    for arg in args:
-        value = frame.FindVariable(arg)
-        result.AppendMessage(
-            bson_as_json(value,
-                         debugger,
-                         verbose=options.verbose,
-                         oneline=options.oneline,
-                         raw=options.raw))
+    def get_cached(self, addr: int) -> bytes:
+        """
+        Find and return a chunk of memory that was previously read from the
+        process. The ``addr`` must be an address within a BSON document that
+        was previously read by a DocumentDisplay.
 
+        The returned buffer begins at exactly ``addr``, and contains the full
+        remainder of the cached segment, which may be much longer than the
+        caller actually needs. We don't need additional granularity for our
+        purposes, though.
+        """
+        segment = self.segment_containing(addr)
+        if not segment:
+            # Memory does not exist?
+            print(f"lldb_bson: Note: Attempted read of uncached address 0x{addr:x}")
+            print("".join(traceback.format_stack()))
+            return b"\0" * 512
+        base_addr, data = segment
+        inner_offset = addr - base_addr
+        return data[inner_offset:]
 
-def bson_type_summary(value, internal_dict):
-    return bson_as_json(value, lldb.debugger)
+    def segment_containing(self, addr: int) -> tuple[int, bytes] | None:
+        """Find the segment that contains ``addr``, or returns None"""
+        for base_addr, data in self._segments.items():
+            end_addr = base_addr + len(data)
+            if addr > end_addr or addr < base_addr:
+                continue
+            return base_addr, data
+        return None
 
+    def read(self, val: SBValue) -> tuple[int, bytes]:
+        """
+        Read the memory segment referenced by the given SBValue
+        """
+        return self.read_at(val.process, val.load_addr, val.size)
 
-def __lldb_init_module(debugger, internal_dict):
-    debugger.HandleCommand(
-        'type summary add -F lldb_bson.bson_type_summary bson_t')
+    def read_at(self, proc: SBProcess, addr: int, size: int) -> tuple[int, bytes]:
+        """
+        Read from ``proc`` a chunk of memory beginning at ``addr`` and continuing
+        for ``size`` bytes.
+        """
+        err = SBError()
+        buf = proc.ReadMemory(addr, size, err)
+        if err.fail:
+            raise LookupError(err.description)
+        self._segments[addr] = buf
+        return addr, buf
 
-    debugger.HandleCommand(
-        'command script add --help \"%s\"'
-        ' -f lldb_bson.bson_as_json_command bson' %
```

Comment:
Update the [LLDB documentation](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/doc/debugging.rst#lldb), since it may no longer be correct.